src/container_runner.py

```diff
@@ -0,0 +1,292 @@
+"""Control a OCI image running via Docker on a host system. Provides a ContainerRunner class."""
+
+import logging
+import subprocess
+import time
+import os
+from typing import Iterable, Optional
+
+from charms.operator_libs_linux.v1 import snap
+
+# Configure logging level based on environment variable or default to INFO
+log_level = os.getenv("LOG_LEVEL", "WARNING").upper()
+
+# Set up the logger
+logger = logging.getLogger(__name__)
+logger.setLevel(log_level)
+
+
+class _Docker:
+    """Private class for handling the installed Docker snap."""
+
+    @property
+    def _docker(self):
+        """Return a representation of the Docker snap."""
+        cache = snap.SnapCache()
+        return cache["docker"]
+
+    def install(self):
+        # Install Docker
+        try:
+            self._docker.ensure(snap.SnapState.Latest, channel="stable")
+            snap.hold_refresh()
+            self._wait_for_docker()
+        except snap.SnapError as e:
+            logger.error("could not install docker. Reason: %s", e.message)
+            logger.debug(e, exc_info=True)
+            raise e
+
+    def _wait_for_docker(self, retries: int = 10, delay: int = 3):
+        """Wait for Docker daemon to be ready."""
+        for _ in range(retries):
+            try:
+                subprocess.check_output(["docker", "info"], universal_newlines=True)
+                logger.info("Docker daemon is ready")
+                return
+            except subprocess.CalledProcessError:
+                logger.warning("Docker daemon is not ready, retrying...")
+                time.sleep(delay)
+        raise Exception("Docker daemon did not become ready in time")
+
+    def _run_command(self, command: str, args: Optional[Iterable[str]] = None) -> str:
```

Comment:
I can't remember if there is or not, but if docker supports outputting JSON (or another machine readable format) that would be good to look into for this so that we don't need to use substring matching to parse the output